src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md

```diff
@@ -0,0 +1,116 @@
+require JS glue code and access to the web APIs and can thus run outside the web,
+the [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target may be better
+suited for you. However, [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md)
+only has partial support for the Rust standard library and does not (easily)
+support interop with C/C++ code.
+
+Like Emscripten, the WASI targets [`wasm32-wasip1`](./wasm32-wasip1.md) and
+[`wasm32-wasip2`](./wasm32-wasip2.md) also provide access to the host environment,
+support interop with C/C++ (and other languages), and support most of the Rust
+standard library. While the WASI targets are portable across different hosts
+(web and non-web), WASI has no standard way of accessing web APIs, whereas
+Emscripten has the ability to run arbitrary JS from WASM and access many web APIs.
+If you are only targeting the web and need to access web APIs, the
+`wasm32-unknown-emscripten` target may be preferable.
+
+## Target maintainers
+
+- Hood Chatham, https://github.com/hoodmane
+- Sam Clegg, https://github.com/sbc100
+- Juniper Tyree, https://github.com/juntyr
+
+## Requirements
+
+This target is cross-compiled. The Emscripten compiler toolchain (emsdk) must be
+installed to link Rust-compiled code with Emscripten's libc and other C/C++
+code. Please refer to <https://emscripten.org/docs/getting_started/downloads.html>
+on how to get started with the emsdk.
+
+## Building the target
+
+Building this target can be done by:
+
+* Configure the `wasm32-unknown-emscripten` target to get built.
+* Configure LLD to be built.
+* Ensure the `WebAssembly` target backend is not disabled in LLVM.
+
+These are all controlled through `config.toml` options. It should be possible
+to build this target on any platform.
+
+## Building Rust programs
+
+Rust programs can be compiled by adding this target via rustup:
+
+```sh
+$ rustup target add wasm32-unknown-emscripten
+```
+
+and then compiling with the target:
+
+```sh
+$ rustc foo.rs --target wasm32-unknown-emscripten
+$ file foo.wasm
+```
+
+## Cross-compilation
+
+This target can be cross-compiled from any host.
+
+## Emscripten ABI Compatibility
+
+The Emscripten compiler toolchain does not follow a semantic versioning scheme
+that clearly indicates when breaking changes to the ABI can be made. If you are
+using the pre-compiled `std` from rustup, it may be compiled with a different
+Emscripten version than the local Emscripten you have installed. If a breaking
+change to the ABI was made in between the two versions, your code will exhibit
+undefined behaviour. If you come across issues, you can rebuild the Rust standard
+library with your local Emscripten version using
+
+```sh
+$ cargo +nightly -Zbuild-std build
+```
+
+## Testing
+
+This target is not extensively tested in CI for the rust-lang/rust repository.
+It's recommended to test the [`wasm32-wasip1`](./wasm32-wasip1.md) target instead
+for WebAssembly compatibility.
```

Comment:
Could you elaborate how contributors can test this target locally?

i.e. Tier 2 target policy says

> The target must provide documentation for the Rust community explaining [..] how to run tests for the target. If at all possible, this documentation should show how to run Rust programs and tests for the target using emulation, to allow anyone to do so. If the target cannot be feasibly emulated, the documentation should explain how to obtain and work with physical hardware, cloud systems, or equivalent.

see #131666, where Jubilee found that

> a fifth of the rustc UI test suite to fail on wasm32-unknown-emscripten when trying to run ./x.py test --target wasm32-unknown-emscripten

Reply:
I was able to get past the test helpers stage with <https://emscripten.org/docs/getting_started/downloads.html> including to clone that repo, use `./emsdk` to fetch and activate the latest emcc toolchain, then use a provided shell script to setup the relevant env vars and make them available to bootstrap.

stage1 std seems to be wonky in the bootstrap test setup

```
warning: dropping unsupported crate type `dylib` for target `wasm32-unknown-emscripten`

warning: `std` (lib) generated 1 warning
```

ui tests failed due to linker-flavors and such and I didn't bother investigating.

Reply:
@hoodmane Since you have gotten most tests to run, could you suggest some text?

Reply:
I've been running tests with:
```sh
./x.py test --target wasm32-unknown-emscripten --skip src/tools/linkchecker
```

Reply:
But you need `emcc` and `node` on the path first. To get `emcc` installed and on the path:
```sh
git clone https://github.com/emscripten-core/emsdk.git --depth 1
./emsdk/emsdk install 3.1.68
./emsdk/emsdk activate 3.1.68
source ./emsdk/emsdk_env.sh
```
And you can use whatever normal way to install node, but if we're going to recommend something, nvm is quite nice:
https://github.com/nvm-sh/nvm#install--update-script

Reply:
Thanks! I've added those instructions